MM-33855 Ad-hoc Tasks

crspeller opened the pull request:
#### Summary
Allows editing of the incident task list. Supports add, remove, edit, and reorder operations.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-33855

agarciamontoro commented on server/incident/service.go:
```diff
@@ -930,16 +983,34 @@ func (s *ServiceImpl) GetChecklistAutocomplete(incidentID string) ([]model.Autoc
 }
 
+// GetChecklistAutocomplete returns the list of checklist items for incidentID to be used in autocomplete
+func (s *ServiceImpl) GetChecklistAutocompleteItem(incidentID string) ([]model.AutocompleteListItem, error) {
```
Also, are the names swapped? This is returning the checklists, so I *think* this should be `GetChecklistAutocomplete`. Maybe I'm missing something?

crspeller replied:
Yes you are right. No need to preserve API compatibility so I did mean to flip them around.

agarciamontoro commented on tests-e2e/cypress/integration/frontstage/incident_rhs_checklist_spec.js:
```diff
@@ -189,6 +189,24 @@ describe('incident rhs checklist', () => {
+
+            cy.addNewTaskFromPostMenu(newTasktext);
+
+            // CHeck that it was created
```
Nit
```suggestion
            // Check that it was created
```

agarciamontoro commented on tests-e2e/cypress/integration/frontstage/incident_rhs_checklist_spec.js:
```diff
@@ -189,6 +189,24 @@ describe('incident rhs checklist', () => {
+
+            cy.executeSlashCommand(`/incident checkadd 0 ${newTasktext}`);
+
+            // CHeck that it was created
```
Nit
```suggestion
            // Check that it was created
```

agarciamontoro commented on tests-e2e/cypress/integration/frontstage/incident_rhs_checklist_spec.js:
```diff
@@ -189,6 +189,24 @@ describe('incident rhs checklist', () => {
+            cy.addNewTaskFromPostMenu(newTasktext);
+
+            // CHeck that it was created
+            cy.findByText(newTasktext).should('exist');
```
Should we isolate this at least to the RHS? Or maybe make the text unique with a timestamp or something similar. Same comment in the other test.

crspeller replied:
Added a timestamp to make the text more unique.

agarciamontoro commented on tests-e2e/cypress/support/plugin_ui_commands.js:
```diff
@@ -65,6 +65,18 @@ Cypress.Commands.add('startIncidentFromRHS', (playbookName, incidentName) => {
+    cy.findByTestId('nameinput').type(taskname);
+
+    // Submit the dialog
+    cy.findByText('Add Task').click();
```
Same isolation comment here. Maybe this selector is too flexible. Is there a way to isolate it? Getting the dialog first and then finding by text, maybe?

crspeller replied:
But this is nice, because it only tests what we care about. It's somewhere on the screen. So if we re-arrange the UI, this test doesn't break :D

agarciamontoro replied:
That's true! I agree that in this case, it's probably more probable that we re-arrange the UI than that someone else adds a "Add Task" text somewhere.

agarciamontoro commented on webapp/src/components/assets/inputs.tsx:
```diff
@@ -0,0 +1,36 @@
+import styled from 'styled-components';
+
+export const BaseInput = styled.input`
+    -webkit-transition: border-color ease-in-out .15s, -webkit-box-shadow ease-in-out .15s;
+    transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s, -webkit-box-shadow ease-in-out .15s;
```
According to [the docs](https://styled-components.com/docs/basics#motivation), we should not need the prefixes here: "Automatic vendor prefixing: write your CSS to the current standard and let styled-components handle the rest." Do we need it here?

crspeller replied:
Probably not. Didn't write this CSS just copy pasted. Willing to horribly break safari if you are 😆

agarciamontoro replied:
I am :man_shrugging:

agarciamontoro commented on webapp/src/components/checklist_item.tsx:
```diff
@@ -355,15 +361,30 @@ export const ChecklistItemDetails = (props: ChecklistItemDetailsProps): React.Re
+                onMouseLeave={() => setShowMenu(false)}
+                data-testid='checkbox-item-container'
+            >
+                <CheckboxContainer >
```
Nit
```suggestion
                <CheckboxContainer>
```

agarciamontoro commented on webapp/src/components/checklist_item_input.tsx:
```diff
@@ -0,0 +1,199 @@
+    const [focus, setFocus] = useState(false);
+    const ref = useRef(null);
+
+    useEffect(() => {
+        if (focus && ref && ref.current) {
+            // @ts-ignore
+            ref.current.focus();
+            setFocus(false);
+        }
+    }, [focus, setFocus]);
```
Not sure what this does

agarciamontoro commented on webapp/src/components/command_input.tsx:
```diff
@@ -0,0 +1,126 @@
+// @ts-ignore
+const AutocompleteTextbox = window.Components.Textbox;
+
+const OverrideWebappStyle = createGlobalStyle`
```
How do the `createGlobalStyle` work? Isn't it the same as writing a CSS file?

crspeller replied:
Pretty much.

agarciamontoro commented on webapp/src/components/widgets/generic_modal.tsx:
```diff
@@ -0,0 +1,135 @@
+    show: boolean;
+}
+
+export default class GenericModal extends React.PureComponent<Props, State> {
```
Why is this a class component and not a functional one?

cpoile replied:
Probably because it's from webapp, right?

crspeller replied:
Yes! Sorry I skipped this on the first pass. I just copy pasted this from the webapp.

agarciamontoro reviewed (changes requested):
Wow, this is a great job, thank you! I'm happy we got these features back :tada:

I left a bunch of comments, a lot of them nits. I also have a couple of UI considerations:
- The cursor is not a pointer on the menu with the drag/edit/delete/assignee buttons. I don't know if it was like that before this PR, but I guess it's easily fixable here in any case.
- I cannot drag items between checklists. I think this is expected, as there is an Add new task button in each checklist, but it felt weird on the first interaction. Something to evaluate on the UX side, probably not on this PR.
- The slash command autocomplete in the edit modal (which is really cool!) has a couple of bugs (I'm on Firefox, I have not tested in other browsers): the whole modal is misaligned, and the spacing in the individual items seems weird. The screenshot below shows both errors:
![image](https://user-images.githubusercontent.com/3924815/112461573-56f1dc00-8d60-11eb-990f-0ba0cde2332a.png)

Also, what happened with the `package.json`? I must admit I was a bit scared when I saw the + 23,299 on the lines changed count :joy:

crspeller commented on server/api/incidents.go:
```diff
@@ -793,7 +795,31 @@ func (h *IncidentHandler) removeTimelineEvent(w http.ResponseWriter, r *http.Req
+	}
+
+	if err = permissions.ViewIncident(userID, incidentID, h.pluginAPI); err != nil {
+		HandleErrorWithCode(w, http.StatusForbidden, "user does not have permissions", nil)
```
Copy pasted error. Fixed in the other function too.

crspeller commented on server/api/incidents.go:
```diff
@@ -941,6 +967,61 @@ func (h *IncidentHandler) addChecklistItem(w http.ResponseWriter, r *http.Reques
+
+	request := model.SubmitDialogRequestFromJson(r.Body)
+	if request == nil {
+		HandleErrorWithCode(w, http.StatusBadRequest, "failed to decode SubmitDialogRequest", nil)
```
But there is no err here?

agarciamontoro replied:
Indeed, I commented on the wrong line. But you fixed it on the one I actually wanted to comment on :joy:

crspeller commented on server/api/incidents.go:
```diff
@@ -941,6 +967,61 @@ func (h *IncidentHandler) addChecklistItem(w http.ResponseWriter, r *http.Reques
+		return
+	}
+
+	if err := h.incidentService.AddChecklistItem(incidentID, userID, checklistNum, checklistItem); err != nil {
```
Nope. It's done in a router `Use` (check `incidentRouterAuthorized`)

agarciamontoro replied:
Oh yeah, that's true, always forget about that. Thanks!

crspeller commented on server/command/command.go:
```diff
@@ -60,7 +62,7 @@ func getCommand(addTestCommands bool) *model.Command {
 func getAutocompleteData(addTestCommands bool) *model.AutocompleteData {
 	slashIncident := model.NewAutocompleteData("incident", "[command]",
-		"Available commands: start, end, update, restart, check, announce, list, commander, info, timeline")
+		"Available commands: start, end, update, restart, check, item, announce, list, commander, info, timeline")
```
The fact this isn't automatic makes me angry.

crspeller commented:
@agarciamontoro 
- Fixed the pointer thing (it was actually a problem before)
- I know about the not being able to drag between checklists. Going to add a separate ticket for that.
- Fixed the strange spacing, but the offset appears intentional? It works the same in the backstage and in the webapp. 
- Not sure why the package-lock.json keeps doing this... I didn't even change anything!

cpoile reviewed (approved):
noice. Can't see anything blocking. Thanks @crspeller!

agarciamontoro commented on server/incident/incident.go:
```diff
@@ -314,13 +317,16 @@ type Service interface {
 	// MoveChecklistItem moves a checklist item from one position to anouther
 	MoveChecklistItem(incidentID, userID string, checklistNumber int, itemNumber int, newLocation int) error
 
 	// GetChecklistAutocomplete returns the list of checklist items for incidentID to be used in autocomplete
```
Still nit:
```suggestion
	// GetChecklistItemAutocomplete returns the list of checklist items for incidentID to be used in autocomplete
```

agarciamontoro reviewed (approved):
Cool! Nothing major blocking on the second review, thank you!

About the `package-lock.json` changes: is it possible that you are using npm 7? Your change is now using the new [lockfile format](https://github.blog/2021-02-02-npm-7-is-now-generally-available/#changes-to-the-lockfile). It seems to be backwards compatible, so I'm ok leaving it as it is. @cpoile and I may have to update, though?
